libs/compact-map/src/Data/Compact/KeyMap.hs

```diff
@@ -248,67 +297,86 @@ insertWithKey' n0 combine path k x kmap = go n0 kmap
                   else Two bmap x0 st'
       where
         i = indexFromSegment bmap j
-        j = path !! n
-    go n t@(Full arr) =
+        j = path VP.! n
+    go !n t@(Full arr) =
       let !st = index arr i
           !st' = go (n + 1) st
        in if st' `ptrEq` st
             then t
             else Full (update arr i st')
       where
         i = indexFromSegment fullNodeMask j
-        j = path !! n
+        j = path VP.! n
+{-# INLINE insertWithKey' #-}
 
 twoLeaf :: Path -> KeyMap v -> Path -> Key -> v -> KeyMap v
-twoLeaf [] _ _ _ _ = error ("the path ran out of segments in twoLeaf case 1.")
-twoLeaf _ _ [] _ _ = error ("the path ran out of segments in twoLeaf case 1.")
-twoLeaf (i : is) leaf1 (j : js) k2 v2
-  | i == j = One i (twoLeaf is leaf1 js k2 v2)
-  | otherwise =
-    if i < j
-      then Two (setBits [i, j]) leaf1 (Leaf k2 v2)
-      else Two (setBits [i, j]) (Leaf k2 v2) leaf1
+twoLeaf path1 leaf1 path2 k2 v2 = go path1 path2
+  where
+    leaf2 = Leaf k2 v2
+    go p1 p2
+      | Just (i, is) <- VP.uncons p1,
+        Just (j, js) <- VP.uncons p2 =
+        if i == j
```

Comment:
I understand your reasoning, but I was using benchmarks as a guide, rather than logical reasoning. In the end what we think the performance should be is not necessarily the performance that we'll get. Using lists for Path makes the worst case 5 times slower than when using primitive vector.

We can certainly discuss it in more detail. If you can come up with the worst case example where vector would perform worse that would be tremendously useful, because then we could benchmark it and reason about the performance hit if any. One way or another `insert` is one of the most important operations in this map, and that is the one that I've optimized.